src/ripple/app/data/SociDB.cpp

```diff
@@ -140,84 +148,99 @@ void convert(soci::blob& from, std::vector<std::uint8_t>& to)
+    convert (from, tmp);
+    to.assign (tmp.begin (), tmp.end());
 
 }
 
-void convert(std::vector<std::uint8_t> const& from, soci::blob& to)
+void convert (std::vector<std::uint8_t> const& from, soci::blob& to)
 {
     if (!from.empty ())
         to.write (0, reinterpret_cast<char const*>(&from[0]), from.size ());
 }
 
-int SqliteWALHook (void* s,
-                   sqlite_api::sqlite3*,
-                   const char* dbName,
-                   int walSize)
+namespace {
+
+/** Run a thread to checkpoint the write ahead log (wal) for
+    the given soci::session every 1000 pages. This is only implemented
+    for sqlite databases.
+
+    Note: According to: https://www.sqlite.org/wal.html#ckpt this
+    is the default behavior of sqlite. We may be able to remove this
+    class.
+*/
+class WALCheckpointer : public Checkpointer, private beast::Thread
 {
-    (reinterpret_cast<WALCheckpointer*>(s))->doHook (dbName, walSize);
-    return SQLITE_OK;
-}
+public:
+    using Connection = sqlite_api::sqlite3;
+
+    WALCheckpointer (sqlite_api::sqlite3&, JobQueue&);
+    ~WALCheckpointer () override;
+
+    static
+    int sqliteWALHook (void* s, sqlite_api::sqlite3*,
+                       const char* dbName, int walSize);
 
-WALCheckpointer::WALCheckpointer (std::shared_ptr<soci::session> const& s,
-                                  JobQueue* q)
-    : Thread ("sqlitedb"), session_(s)
+private:
+    void runCheckpoint (const char* db, int walSize);
+    void run ();
+    void checkpoint ();
+
+    using LockType = std::mutex;
+    using ScopedLockType = std::lock_guard<LockType>;
+
+    sqlite_api::sqlite3& conn_;
+    LockType mutex_;
+    JobQueue& jobQueue_;
+    bool running_ = false;
+
+};
+
+int WALCheckpointer::sqliteWALHook (
+    void* cp, sqlite_api::sqlite3*, const char* dbName, int walSize)
 {
-    if (auto session =
-            dynamic_cast<soci::sqlite3_session_backend*>(s->get_backend ()))
-        conn_ = session->conn_;
+    if (auto checkpointer = reinterpret_cast <WALCheckpointer*> (cp))
+    {
+        checkpointer->runCheckpoint (dbName, walSize);
+        return SQLITE_OK;
+    }
+    throw std::logic_error ("Didn't get a WALCheckpointer");
+}
 
-    if (!conn_) return;
+WALCheckpointer::WALCheckpointer (sqlite_api::sqlite3& conn, JobQueue& q)
+        : Thread ("sqlitedb"), conn_ (conn), jobQueue_ (q)
+{
     startThread ();
-    setupCheckpointing (q);
+    sqlite_api::sqlite3_wal_hook (&conn_, &sqliteWALHook, this);
 }
 
 WALCheckpointer::~WALCheckpointer ()
 {
-    if (!conn_) return;
     stopThread ();
 }
 
-void WALCheckpointer::setupCheckpointing (JobQueue* q)
-{
-    if (!conn_) return;
-    q_ = q;
-    sqlite_api::sqlite3_wal_hook (conn_, SqliteWALHook, this);
-}
-
-void WALCheckpointer::doHook (const char* db, int pages)
+void WALCheckpointer::runCheckpoint (const char* db, int pages)
 {
-    if (!conn_) return;
-
-    if (pages < 1000)
+    if ((pages + 1) % checkpointPageCount)
```

Comment:
I don't believe this is correct. This needs to change back to the original `if (pages < checkpointPageCount)`. I left a comment in a previous commit explaining why (pages is the size of the wal file, which is consumed in a checkpoint).